src/waylandhelper.cpp

```diff
@@ -0,0 +1,46 @@
+#include "waylandhelper.h"
+
+WaylandHelper::WaylandHelper(QWindow *parent)
+    : QObject(parent)
+    , surface(Surface::fromWindow(parent))
+    , registry(new Registry(this))
+    , connection(ConnectionThread::fromApplication(this))
+{
+    if (!surface->isValid())
+        qFatal("Invalid Wayland surface");
+    registry.create(connection);
+    registry.setup();
+    if (!registry.isValid())
+        qFatal("Invalid Wayland registry");
+    connect(connection, &ConnectionThread::connectionDied, &registry, &Registry::destroy);
+    connect(&registry, &Registry::exporterUnstableV2Announced, this, &WaylandHelper::onXdgExporterAnnounced);
+    connect(&registry, &Registry::interfacesAnnounced, this, &WaylandHelper::onInterfacesAnnounced);
+}
+
+void WaylandHelper::onXdgExporterAnnounced(uint name, uint version)
+{
+    xdgExporter = registry.createXdgExporter(name, version, this);
+    if (!xdgExporter->isValid())
+        qFatal("Invalid Wayland XdgExporter");
+    xdgExportedTopLevel = xdgExporter->exportTopLevel(surface, this);
```

Comment:
What if we store inside the class only `xdgExportedTopLevel` and check for it inside `onInterfacesAnnounced` instead?

Reply:
Honestly, I'm not sure. Nothing in the docs says that must keep `XdgExporter` around while `XdgExported` is used, but it feels safer this way.

Reply:
You don't keep the exact type. It just a pointer which will be destroyed with the parent, so I think we could skip it.

Reply:
Ok, I'll try that.

Reply:
Seems to work this way, done.